### ISSUE TITLE
Calculator: re-write

### ISSUE THREAD

**Issue by GuiltyDolphin** [duckduckgo/zeroclickinfo-goodies#1880]
_(I blabber, scroll down a bit if you want to see some issues and suggestions)_

_Disclamer: I'm not a long-time DuckDuckHack contributer, I have a lot to learn and I'm not as well versed in Perl as many of you will be (I mostly write Haskell), so please, feel free to scream (or politely shout) at me if I've made some horrible error in the way things are supposed to work._

**_Begin**_

I recently wrote a small fix for the Calculator Goodie, but had a bit of pain finding where exactly the issue was due to the verbosity of the goodie. @moollaza informed me that the Calculator is one of the oldest IAs, so through updates it is getting a bit messy.

Whils't doing my fix, I noticed some odities in the way the goodie was written, and since then I have looked further in to it and come up with many suggestions for improvements.

**I propose that the Calculator IA be re-written**

`Calculator.pm` is approximately 190 sloc, with roughly 120 tests - not huge. I believe that a re-write would allow this code-base to be reduced significantly, whilst making the goodie more robust and potentially add more features.

I am aware that there are other powerful calculator implementations, such as [this](https://duck.co/ia/view/symbolab) spice, and [this](https://github.com/duckduckgo/zeroclickinfo-spice/pull/1727) promising GUI implementation; but Perl is a very powerful language, and a **lot** of people have invested in it - resulting in [some powerful math libraries](https://metacpan.org/search?q=Math).

The current Calculator implementation is very focused on using regular expressions to attempt to normalize user queries (`s/ (\d+?)E(-?\d+)([^\d]|\b) /\($1 * 10**$2\)$3/ixg` `2e3` anyone?) many of which are unnecessary.

I propose that a 'new' Calculator Goodie be made, with the intent of giving the user as much mathematical power as possible, whilst keeping the codebase clean and well documented, and delegating hard-work to libraries such as `Math::BigFloat` and `Math::BigInt`.
## Some current issues

_I would include screenshots... but think of the scrolling..._
1. _The code._
   It ain't pretty... And I'm pretty sure a lot of it doesn't need to be there. A good re-write would naturally fix this through careful planning and constant refactoring.
2. _Queries._ Some queries **just don't work**. Take for example, `e` (`* 10 to a power`). The query `1 + 3e8` works fine, but `3e8` on its own? Nope. This could be taken as erring on the side of safety, who knows what `3e8` could really be referring to (things like `2e3` and `2e4` already come up with IA results)? But prefixing the query with `calculate` doesn't fix this, even though the user is being _very_ explicit. This should change.
3. _Premature delegation._ The [spice](https://duck.co/ia/view/symbolab) mentioned earlier has its uses, but they lie with more complex mathematics such as calculus. The current Goodie is relying on it too much. As an example, the query `calculate 1 + 2e`. This isn't a complicated computation, and should be handled directly by the Goodie, rather than having to send out an API call (as it currently does). Besides, the SymboLab call is (by default, don't know if you can change this) only giving back 5 decimal digits! Perl supports 40 (ish) **_cached**_ digits of `e` out-of-the-box with `Math::BigFloat->bexp(1)`.

I have many other issues and things that could be improved noted down, but I'll stop here for now, hopefully I've convinced you.

If you agree - I'll be happy to put time into this, I think this is a very fundamental Goodie (too much maths?).
## TL;DR

_(you lazy... - that took me ages to write!)_

**I think the calculator needs re-writing.**

------ 
IA Page:  https://duck.co/ia/view/calculator

**Comment by mintsoft** [comment]
@GuiltyDolphin Hi Ben

You're definitely correct that the Calculator is probably the most difficult IA to work on currently. I spent some time looking at other options about a year ago and came to the conclusion that the main problem difficulty stems from us handing the hard-work over to Perl itself (https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Calculator.pm#L128) meaning that we have to do a lot of sanitisation in order to protect us from Perl injection attacks.

Essentially I think to make everything easier/more consistent we need to build/use a grammar/expression engine. Marpa (https://metacpan.org/pod/distribution/Marpa-R2/pod/Marpa_R2.pod) is an option, which we use in https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/BinaryLogic.pm#L38, however it may be a little too generic.

There are some calculator modules on cpan already: https://metacpan.org/search?q=calculator however I can't see one that will just do what we want. 

I'd be happy to work with you on a rework of the calculator if you want some perl/community guidance?

**Comment by GuiltyDolphin** [comment]
@mintsoft Having a fixed grammar would definitely help prevent injection attacks - I've recently spent a fair amount of time working with parsers (in Haskell though). I can see the advantage of just being able to add new terms and the relevant routines when implementing new features.

[This](http://perl-begin.org/uses/text-parsing/#advanced-parsing) page links to a few parser-generators, I'll have a look into them, but to be honest, Marpa looks pretty good.

I'd rather steer clear of existing calculator implementations as the logic shouldn't be terribly complicated to implement and I think it'd be much easier to extend if the main source is directly available.

I think another main issue is with triggers. Some things should obviously trigger it like `1 + 7`, but things like `2dozen` are less obvious and I hope for some community input as to preferred triggers.

Maybe the last point will be less of an issue if using a grammar parser, as it may better be able to dictate which terms should trigger the calculator.

Yes, I accept your offer! I'll need to ping quite a lot of ideas for this makeover.

**Comment by mintsoft** [comment]
Excellent, I'm on slack and on here if you want to discuss anything, feel free to just ping me with ideas etc!

I don't think triggering should be too hard, we can easily approximate something that 'looks like' something calculable, i.e. contains numbers and operators or functions ; it'll probably end up a fairly sizable regex. However triggers don't have to be exact, just as long as they're not unnecessarily broad.

**Comment by mintsoft** [comment]
Actually this looks pretty straight forward: http://search.cpan.org/~hakanardo/Math-Expr-0.4/Expr.pm

**Comment by GuiltyDolphin** [comment]
Hmm.. Yeah, I saw that, but it doesn't seem to do much. Just parse the basic operations into a tree. It is... Odd. I can't find a way to actually evaluate or even retrieve expressions (the `Simplify` sub?), and it is very lax with its parsing. I'm not sure this'd be the best bet (let me know if you find out how to use it). It mentions using some databases containing terms, but I can't find anything more on it.

**Comment by mintsoft** [comment]
I think the intention around that module is that the execution is down to the consumer of the tree, I'd guess that a depth-first traversal would execute in BODMAS order

**Comment by GuiltyDolphin** [comment]
Looking at Marpa it seems very promising, and it has thorough documentation. Mind taking a look at [this](https://stackoverflow.com/questions/33630324/is-it-possible-to-use-perls-marpa-parser-for-a-public-network-server) thread on unsafe input? I don't think it will be an issue (or any more of an issue than alternatives) for a well-defined grammar, but better safe than sorry.

**Comment by mintsoft** [comment]
I don't perceive it being a problem as long as the grammar is well defined as you say; it is probably worth us testing it with some pathological inputs within the test cases (odd unicode characters, null bytes etc) to ensure that it behaves sensibly. Judging by Jeff's reply (the author of marpa), it's just to essentially cover all possibilities.

**Comment by GuiltyDolphin** [comment]
The old calculator Goodie went out of it way to normalize the user input for re-display to the user; to what extent will this be required by the new Calculator? Of course removing duplicate whitespace and very easy operations can be done, but I'd rather not have to write a bunch of regexes just to do effectively the same parsing as Marpa.

I'm still looking into whether I can do this easily with Marpa at the same time as producing the valued output.

**Comment by mintsoft** [comment]
I think the output when driven from the parse tree would be sufficient, although it's difficult to say without seeing one obviously.

**Comment by GuiltyDolphin** [comment]
Using something similar to [this](https://metacpan.org/pod/distribution/Marpa-R2/pod/Scanless/DSL.pod#Synopsis), I'm able to implement two functions for each action - one that generates a pretty-print syntax, and one that generates the value.

Could then use the pretty-print version for the input display, and the actual value in the result.

**Comment by mintsoft** [comment]
That sounds like it could work out; although I guess we'd need to parse the expression twice?

**Comment by mintsoft** [comment]
No we wouldn't, the object that represents the operator can just have two methods!

**Comment by GuiltyDolphin** [comment]
Yeah, the actual parsing happens once to generate the tree, then the individual methods can be called for generating the input string and output.

**Comment by GuiltyDolphin** [comment]
Yikes... The old calculator gives `256` as the result (and it is even in the tests!) to `2 squared ^ 3`, which is `4 ^ 3 = 64`. I was wondering why that test wasn't passing (new calc)!

**Comment by mintsoft** [comment]
Hmm, I don't know about that too much; if it was expressed as `2^2^3` then it's the output I'd expect. What's the difference between expressing it as `2 squared` or `2^2`?

**Comment by GuiltyDolphin** [comment]
I would read it as 2 squared to the power of 3, but the answer it was giving was `2 to the power of (squared) to the power of 3` - it was substituting `2 squared` for `2 ^ 2` rather than `4`, so the precedence was getting mucked up.

**Comment by GuiltyDolphin** [comment]
@mintsoft Do you think `squared` should be more or less binding than `^`? More binding: `2 ^ 3 squared = 512`, less binding `2 ^ 3 squared = 64`.

**Comment by mintsoft** [comment]
Good question, I guess to me it's ambiguous. Thinking about it, I don't think people would tend to refer to "squared" as an operator being applied to the output of an expression, only to a number directly so I would expect `2 ^ 3 squared` to be the same as `2^(3^2)` as the squared applies only to the 3. That said, neither way around feels 'right'; if it was said to me in a conversation I'd ask for clarification!

**Comment by GuiltyDolphin** [comment]
@mintsoft I've managed to get it to pass all the tests now (had to modify some tests that were actually incorrect - old calculator was producing some incorrect values for trig functions). Taken a lot longer than anticipated because of some issues with phone numbers, weird issue with the `Math::Complex` library trying to convert numbers (and a couple of other things). Just need to clean it up and rebase my commits into something sensible then make a PR.

**Comment by mintsoft** [comment]
Excellent! Just ping me in the PR and I'll take a look!

**Comment by GuiltyDolphin** [comment]
@mintsoft Regarding the old triggers, I'm not sure how much these are needed in the new calculator, as if the string cannot be parsed by the grammar then the calculator won't accept the input. Additionally the old trigger regex is quite restrictive on what functions/operators will trigger the response. What would your recommendation be? Is it fine to allow the calculator to attempt to parse each input?

**Comment by mintsoft** [comment]
@GuiltyDolphin I think it would be unwise to have it trigger for every query. Just a very loose regex trigger will suffice, which we can then validate with the grammar

**Comment by fzzr-** [comment]
issue #3450 about expressions starting with `-`.

**Comment by moollaza** [comment]
This is now being replaced by #4020
